Set up preference page global to all plugins

chanseokoh opened the pull request:
Sets up a new plugin bundle to accommodate plugin-wide preferences. Also a pre-requisite for #138. It adds a root entry "Google Cloud Tools" to the preference page. (See the screenshot below.)

This bundle is a plugin (i.e., contains 'plugin.xml') and also has an Activator. I thought these are necessary to set up a preference store and UI.

![preference_page](https://cloud.githubusercontent.com/assets/10523105/16062528/a3740cce-3261-11e6-99d9-5734b3e3323f.png)

I still have to find a way to launch a web browser when a user clicks on the hyperlink (the "Google Privacy Policy" text).

About `AnalyticsOptInFieldEditor`: I created this class in order to put two controls (`BooleanFieldEditor` and `Link`) into a `Group` (i.e., a panel with a border) to have a nice-looking UI. I couldn't find a way to do so without directly extending the general base `FieldEditor` class.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/messages.properties:
```diff
@@ -0,0 +1,3 @@
+FIELD_EDITOR_ANALYTICS_DISCLAIMER=Your use of this plugin is subject to the Apache 2.0 software license and the <a href="privacy-policy">Google Privacy Policy</a>.
+FIELD_EDITOR_ANALYTICS_GROUP_TITLE=Usage Statistics
+FIELD_EDITOR_ANALYTICS_OPT_IN_TEXT=&Share anonymous usage statistics of Google Cloud Plugins with Google
```
The standard Eclipse term is _plug-ins_ with a hyphen. Bizarre but it is what it is.

That said, perhaps we could just use _Google Cloud Tools_ and avoid the issue entirely :-)

briandealwis commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/messages.properties:
```diff
@@ -0,0 +1,3 @@
+FIELD_EDITOR_ANALYTICS_DISCLAIMER=Your use of this plugin is subject to the Apache 2.0 software license and the <a href="privacy-policy">Google Privacy Policy</a>.
```
I think it's worth putting the link in here too.  Since this would be translated, the href should also point to the appropriate language text page

chanseokoh replied:
Nice idea (to localize the URL).

briandealwis commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/AnalyticsOptInFieldEditor.java:
```diff
@@ -0,0 +1,73 @@
+        new BooleanFieldEditor(name, Messages.FIELD_EDITOR_ANALYTICS_OPT_IN_TEXT, group);
+
+    Link link = new Link(group, SWT.NONE);
+    link.setText(Messages.FIELD_EDITOR_ANALYTICS_DISCLAIMER);
```
You need to add a selection listener to the `Link` instance.  `event.text` will be the `href` text.  To open a browser, using the platform browser support (which allows opening an internal browser):
```
IWorkbenchBrowserSupport browserSupport = PlatformUI.getWorkbench().getBrowserSupport();
String browserId = null;
IWebBrowser browser = browserSupport.createBrowser(browserId);
browser.openURL(url);
```

chanseokoh replied:
Thanks a lot for the code! The internal browser works fine.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/AnalyticsOptInFieldEditor.java:
```diff
@@ -0,0 +1,73 @@
+   */
+  public AnalyticsOptInFieldEditor(String name, Composite parent) {
+    group = new Group(parent, SWT.SHADOW_OUT);
+    group.setText(Messages.FIELD_EDITOR_ANALYTICS_GROUP_TITLE);
```
You should configure the `Group`'s layout.  It's likely easiest to do:
`GridLayoutFactory.swtDefault().applyTo(group)`.  That should correct the inter-widget spacing of the widgets within the group.

chanseokoh replied:
Ah, thanks. It's my first time using SWT. I did it in `doFillIntoGrid` (some example did it so), but not sure if this can be done once in the constructor.

briandealwis commented:
The `FieldEditor`s/`FieldEditorPreferencePage` were intended to make creating simple pref pages easy.  In this case you might have just created a custom `PreferencePage`.  But no matter.

I wonder about the bundle name: although it would be good to have a top-level page, it seems odd to have a top-level page include preferences for functionality defined in a different plugin (`.usagetracker`)

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/Activator.java:
```diff
@@ -0,0 +1,61 @@
+  // The shared instance
+  private static Activator plugin;
+
+  /**
```
please remove or expand obvious JavaDoc comments

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/Activator.java:
```diff
@@ -0,0 +1,61 @@
+ */
+public class Activator extends AbstractUIPlugin {
+
+  // The plug-in ID
```
please remove or expand obvious JavaDoc comments

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/Activator.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  /*
+   * (non-Javadoc)
```
please remove or expand obvious JavaDoc comments

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/Activator.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  /*
+   * (non-Javadoc)
```
please remove or expand obvious JavaDoc comments

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/Activator.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  /**
+   * Returns the shared instance
```
period

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/Activator.java:
```diff
@@ -0,0 +1,61 @@
+
+  /**
+   * Returns an image descriptor for the image file at the given
+   * plug-in relative path
```
period

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/Activator.java:
```diff
@@ -0,0 +1,61 @@
+   * Returns an image descriptor for the image file at the given
+   * plug-in relative path
+   *
+   * @param path the path
```
please remove or expand obvious JavaDoc comments

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/CloudToolsPreferencePage.java:
```diff
@@ -0,0 +1,35 @@
+ * This class represents a preference page that is contributed to the
+ * Preferences dialog. By subclassing <samp>FieldEditorPreferencePage</samp>, we
+ * can use the field support built into JFace that allows us to create a page
+ * that is small and knows how to save, restore and apply itself.
```
restore,

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/PreferenceInitializer.java:
```diff
@@ -0,0 +1,21 @@
+
+  /*
+   * (non-Javadoc)
+   * 
```
please remove or expand obvious JavaDoc comments

chanseokoh commented:
@briandealwis About the `usagetracker` setting on the top-level page: I agree it's odd this is on the top level. I will probably push it down to a child node, but I am inclined to do it later because otherwise the top-level page is totally empty. For now, having a child node seems like an unnecessary complexity given that we have only one setting.

chanseokoh commented:
The last commit also introduces a Logger instance.

chanseokoh commented:
PTAL

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/messages.properties:
```diff
@@ -0,0 +1,4 @@
+FIELD_EDITOR_ANALYTICS_DISCLAIMER=Your use of this plug-in is subject to the Apache 2.0 software license and the <a href="privacy-policy">Google Privacy Policy</a>.
```
Why are these all prefixed with FIELD_EDITOR? They feel more general than that. e.g. just ANALYTICS_DISCLAIMER should be fine.

elharo replied:
I don't think we need to mention the license here. We should take it out so the privacy policy is more prominent. What we want is something like "Share anonymous usage statistics of Google Cloud Tools with Google subject to the Google Privacy Policy."

briandealwis replied:
They'll have agreed to the license during installation.

elharo commented:
Any ideas how we might write tests for any of this?

chanseokoh commented:
I removed FIELD_EDITOR_ in the message bundle keys and changed the disclaimer text.

I don't have any good idea about the tests. The provided framework reads preferences from a file and writes them to it directly.

elharo commented on plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/messages.properties:
```diff
@@ -1,4 +1,4 @@
-FIELD_EDITOR_ANALYTICS_GROUP_TITLE=Usage Statistics
-FIELD_EDITOR_ANALYTICS_OPT_IN_TEXT=&Share anonymous usage statistics of Google Cloud Tools with Google
+ANALYTICS_DISCLAIMER=Sharing usage statistics is subject to the <a href="privacy-policy">Google Privacy Policy</a>.
+ANALYTICS_PREF_GROUP_TITLE=Usage Statistics
```
PREF --> PREFERENCE (Google style is not to abbreviate)

elharo commented:
LGTM
